sidechain/state/src/impls.rs

```diff
@@ -0,0 +1,162 @@
+//! Implement the sidechain state traits. This is only sgx-compatible due to [`SgxExternalities`]
+//! missing some features in `std`.
+
+use crate::{sgx_reexports::*, Error, SidechainDB, SidechainState, StateHash, StateUpdate};
+use codec::{Decode, Encode};
+use itp_storage::keys::storage_value_key;
+use log::error;
+use sgx_externalities::SgxExternalitiesTrait;
+use sp_core::{hashing::blake2_256, H256};
+use sp_std::prelude::Vec;
+
+impl<SB, T> SidechainState for SidechainDB<SB, T>
+where
+	T: SgxExternalitiesTrait + StateHash + Clone,
+	SB: Clone,
+{
+	type Externalities = T;
+	type StateUpdate = StateUpdate;
+	type Hash = H256;
+
+	fn state_hash(&self) -> Self::Hash {
+		self.ext.hash()
+	}
+
+	fn apply_state_update(&mut self, state_payload: &Self::StateUpdate) -> Result<(), Error> {
+		// Todo: how do we ensure that the apriori state hash matches: See #421
+		// ensure!(self.state_hash() == state_payload.state_hash_apriori(), Error::InvalidAprioriHash);
+		let mut state2 = self.ext.clone();
+
+		state2.execute_with(|| {
+			state_payload.state_update.iter().for_each(|(k, v)| {
+				match v {
+					Some(value) => sp_io::storage::set(k, value),
+					None => sp_io::storage::clear(k),
+				};
+			})
+		});
+
+		// Todo: Consequence of #421
+		// ensure!(state2.hash() == state_payload.state_hash_aposteriori(), Error::InvalidStorageDiff);
+		self.ext = state2;
+		self.ext.prune_state_diff();
+		Ok(())
+	}
+
+	fn get_with_name<V: Decode>(&self, module_prefix: &str, storage_prefix: &str) -> Option<V> {
+		let res = self
+			.get(&storage_value_key(module_prefix, storage_prefix))
+			.map(|v| Decode::decode(&mut v.as_slice()))
+			.transpose();
+
+		match res {
+			Ok(res) => res,
+			Err(e) => {
+				error!(
+					"Error decoding storage: {}, {}. Error: {:?}",
+					module_prefix, storage_prefix, e
+				);
+				None
+			},
+		}
+	}
+
+	fn set_with_name<V: Encode>(&mut self, module_prefix: &str, storage_prefix: &str, value: V) {
+		self.set(&storage_value_key(module_prefix, storage_prefix), &value.encode())
+	}
+
+	fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
+		self.ext.get(key).cloned()
+	}
+
+	fn set(&mut self, key: &[u8], value: &[u8]) {
+		// self.ext.insert(key.to_vec(), value.to_vec());
+		self.ext.execute_with(|| sp_io::storage::set(key, value))
+	}
+}
+
+impl<E: SgxExternalitiesTrait + Encode> StateHash for E {
+	fn hash(&self) -> H256 {
+		self.state().using_encoded(blake2_256).into()
+	}
+}
+
+// Most of the tests here can't be run in `std` just because `SgxExternalities` does not implement
+// `Encode` in std. See: https://github.com/integritee-network/sgx-runtime/issues/26
+#[cfg(feature = "test")]
```

Comment:
but are they as part of the enclave-runtime tests?

Reply:
Yes 👍.